cpp/oneapi/dal/test/engine/csr_table_builder.hpp

```diff
@@ -0,0 +1,175 @@
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include "oneapi/dal/table/csr.hpp"
+#include "oneapi/dal/backend/primitives/utils.hpp"
+#include "oneapi/dal/test/engine/math.hpp"
+
+namespace oneapi::dal::test::engine {
+
+namespace pr = dal::backend::primitives;
+
+/**
+* Generates random CSR table based on inputs
+*/
+struct csr_table_builder {
+    using Float = float;
+    std::int64_t row_count_, column_count_;
+    float nonzero_fraction_;
+    sparse_indexing indexing_;
+    const dal::array<Float> data_;
+    const dal::array<std::int64_t> column_indices_;
+    const dal::array<std::int64_t> row_offsets_;
+
+    csr_table_builder(std::int64_t row_count,
+                      std::int64_t column_count,
+                      float nnz_fraction = 0.3,
+                      sparse_indexing indexing = sparse_indexing::one_based)
+            : row_count_(row_count),
+              column_count_(column_count),
+              nonzero_fraction_(nnz_fraction),
+              indexing_(indexing),
+              data_(dal::array<Float>::empty(nnz_fraction * row_count * column_count)),
+              column_indices_(
+                  dal::array<std::int64_t>::empty(nnz_fraction * row_count * column_count)),
+              row_offsets_(dal::array<std::int64_t>::empty(row_count + 1)) {
+        std::int64_t total_count = row_count_ * column_count_;
+        std::int64_t nonzero_count = total_count * nnz_fraction;
+        std::int64_t indexing_shift = bool(indexing == sparse_indexing::one_based);
+
+        std::uint32_t seed = 42;
+        std::mt19937 rng(seed);
+        std::uniform_real_distribution<Float> uniform_data(-10.0f, 10.0f);
+        std::uniform_int_distribution<std::int64_t> uniform_indices(
+            0,
+            column_count_ - 1 - indexing_shift);
+        std::uniform_int_distribution<std::int64_t> uniform_ind_count(1, column_count_ - 2);
+
+        auto data_ptr = data_.get_mutable_data();
+        auto col_indices_ptr = column_indices_.get_mutable_data();
+        auto row_offsets_ptr = row_offsets_.get_mutable_data();
+        // Generate data
+        for (std::int32_t i = 0; i < nonzero_count; ++i) {
+            data_ptr[i] = uniform_data(rng);
+        }
+        // Generate column indices and fill row offsets
+        std::int64_t row_idx = 0;
+        std::int64_t fill_count = 0;
+        row_offsets_ptr[0] = indexing_shift;
+        while (fill_count < nonzero_count && row_idx < row_count_) {
```

Comment:
Another idea how to generate the data:
1. Generate the data with the requested parameters in coordinate format first. I.e. generate the triplets: (row_id, column_id, value) where row_id amd column_id can be generated using integer uniform distribution.
If there are duplications in (row_id, col_id), which should be a rare scenario - continue generation until the requested number of non-zero values are generated. 
2. Convert the data from coordinate format into CSR using sorting.